k2/torch/bin/online_decode.cu

```diff
@@ -0,0 +1,328 @@
+  fbank_opts.frame_opts.dither = 0;
+  fbank_opts.frame_opts.frame_shift_ms = FLAGS_frame_shift_ms;
+  fbank_opts.frame_opts.frame_length_ms = FLAGS_frame_length_ms;
+  fbank_opts.mel_opts.num_bins = FLAGS_num_bins;
+  fbank_opts.device = device;
+
+  kaldifeat::Fbank fbank(fbank_opts);
+
+  K2_LOG(INFO) << "Load wave files";
+  auto wave_data = k2::ReadWave(wave_filenames, FLAGS_sample_rate);
+
+  for (auto &w : wave_data) {
+    w = w.to(device);
+  }
+
+  K2_LOG(INFO) << "Compute features";
+  std::vector<int64_t> num_frames;
+  auto features_vec = k2::ComputeFeatures(fbank, wave_data, &num_frames);
+
+  // Note: math.log(1e-10) is -23.025850929940457
+  auto features = torch::nn::utils::rnn::pad_sequence(features_vec, true,
+                                                      -23.025850929940457f);
+
+  K2_LOG(INFO) << "Load neural network model";
+  torch::jit::script::Module module = torch::jit::load(FLAGS_jit_pt);
+  module.eval();
+  module.to(device);
+
+  int32_t subsampling_factor = module.attr("subsampling_factor").toInt();
+  torch::Dict<std::string, torch::Tensor> sup;
+  sup.insert("sequence_idx", torch::arange(num_waves, torch::kInt));
+  sup.insert("start_frame", torch::zeros({num_waves}, torch::kInt));
+  sup.insert("num_frames",
+             torch::from_blob(num_frames.data(), {num_waves}, torch::kLong)
+                 .to(torch::kInt));
+
+  torch::IValue supervisions(sup);
+
+  std::vector<torch::IValue> inputs;
+  inputs.emplace_back(std::move(features));
+  inputs.emplace_back(supervisions);
+
+  K2_LOG(INFO) << "Compute nnet_output";
+  // the output for module.forward() is a tuple of 3 tensors
+  auto outputs = module.forward(inputs).toTuple();
+  assert(outputs->elements().size() == 3u);
+
+  auto nnet_output = outputs->elements()[0].toTensor();
+  auto memory = outputs->elements()[1].toTensor();
+
+  // memory_key_padding_mask is used in attention decoder rescoring
+  // auto memory_key_padding_mask = outputs->elements()[2].toTensor();
+
+  k2::FsaClass decoding_graph;
+
+  if (FLAGS_use_ctc_decoding) {
+    K2_LOG(INFO) << "Build CTC topo";
+    decoding_graph =
+        k2::CtcTopo(nnet_output.size(2) - 1, /*modified*/ false, device);
+  } else {
+    K2_LOG(INFO) << "Load HLG.pt";
+    decoding_graph = k2::LoadFsa(FLAGS_hlg, device);
+    K2_CHECK(decoding_graph.HasTensorAttr("aux_labels") ||
+             decoding_graph.HasRaggedTensorAttr("aux_labels"));
+  }
+
+  K2_LOG(INFO) << "Decoding";
+
+  auto decoding_fsa = k2::FsaToFsaVec(decoding_graph.fsa);
+  k2::OnlineIntersectDensePruned decoder(
+      decoding_fsa, num_waves, FLAGS_search_beam, FLAGS_output_beam,
+      FLAGS_min_activate_states, FLAGS_max_activate_states);
+
+  std::vector<std::string> texts(num_waves, "");
+
+  int32_t T = nnet_output.size(1);
+  int32_t chunk_size = 20;  // 20 frames per chunk
+  int32_t chunk_num = (T / chunk_size) + ((T % chunk_size) ? 1 : 0);
+
+  for (int32_t c = 0; c < chunk_num; ++c) {
```

Comment:
This demo doen not use streaming acoustic model, just splits the whole nnet_output into chunks here.